From 1429e0303d5651ca0cb1d2be0e93b32eb8ce1386 Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Tue, 29 Mar 2016 11:29:52 -0700
Subject: [PATCH 20/25] Disable AVX/SSE on unwind-c.c and unwind-dw2.c

Since %ecx can't be used for both DRAP register and __builtin_eh_return,
we must avoid stack realignment when __builtin_eh_return is used.  We
disable AVX, when the incoming stack is 16-byte aligned, or SSE, when the
incoming stack is 4-byte aligned, on unwind-c.c and unwind-dw2.c to avoid
DRAP register in 32-bit x86 libgcc.

	PR target/70439
	* config.host (tmake_file): Add i386/32/t-eh-return-no-avx for
	32-bit x86 Linux.  Add i386/32/t-eh-return-no-sse for 32-bit
	x86 Cygwin and Solaris.
	* libgcc/config/i386/32/t-eh-return-no-avx: New file.
	* libgcc/config/i386/32/t-eh-return-no-sse: Likewise.
---
 libgcc/config.host                       | 16 ++++++++++++++++
 libgcc/config/i386/32/t-eh-return-no-avx |  6 ++++++
 libgcc/config/i386/32/t-eh-return-no-sse |  6 ++++++
 3 files changed, 28 insertions(+)
 create mode 100644 libgcc/config/i386/32/t-eh-return-no-avx
 create mode 100644 libgcc/config/i386/32/t-eh-return-no-sse

diff --git a/libgcc/config.host b/libgcc/config.host
index b61a579..f68c60d 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -1386,3 +1386,19 @@ i[34567]86-*-linux* | x86_64-*-linux*)
 	tm_file="${tm_file} i386/value-unwind.h"
 	;;
 esac
+
+case ${host} in
+i[34567]86-*-* | x86_64-*-*)
+  if test "${host_address}" = 32; then
+    case ${host} in
+    *-*-cygwin* | *-*-solaris2*)
+      # Disable SSE on unwind-c.c and unwind-dw2.c to avoid DRAP register.
+      tmake_file="${tmake_file} i386/${host_address}/t-eh-return-no-sse"
+      ;;
+    *)
+      # Disable AVX on unwind-c.c and unwind-dw2.c to avoid DRAP register.
+      tmake_file="${tmake_file} i386/${host_address}/t-eh-return-no-avx"
+      ;;
+    esac
+  fi
+esac
diff --git a/libgcc/config/i386/32/t-eh-return-no-avx b/libgcc/config/i386/32/t-eh-return-no-avx
new file mode 100644
index 0000000..782dd58
--- /dev/null
+++ b/libgcc/config/i386/32/t-eh-return-no-avx
@@ -0,0 +1,6 @@
+# Since %ecx can't be used for both DRAP register and __builtin_eh_return,
+# we must avoid stack realignment when __builtin_eh_return is used.  When
+# the incoming stack is 16-byte aligned, we disable AVX on unwind-c.c and
+# unwind-dw2.c to avoid DRAP register without disabling SSE.
+
+unwind-c.o unwind-c_s.o unwind-dw2.o unwind-dw2_s.o : HOST_LIBGCC2_CFLAGS += -mno-avx
diff --git a/libgcc/config/i386/32/t-eh-return-no-sse b/libgcc/config/i386/32/t-eh-return-no-sse
new file mode 100644
index 0000000..dc9bf06
--- /dev/null
+++ b/libgcc/config/i386/32/t-eh-return-no-sse
@@ -0,0 +1,6 @@
+# Since %ecx can't be used for both DRAP register and __builtin_eh_return,
+# we must avoid stack realignment when __builtin_eh_return is used.  When
+# the incoming stack is 4-byte aligned, we disable SSE on unwind-c.c and
+# unwind-dw2.c to avoid DRAP register.
+
+unwind-c.o unwind-c_s.o unwind-dw2.o unwind-dw2_s.o : HOST_LIBGCC2_CFLAGS += -mno-sse
-- 
2.8.1

